### PR TITLE
Wip/slick3 support

### PR DESCRIPTION
**Pull request by dotta**
This is finally ready for review.

The commit message should provide a nice overview of most changes bundled in this big PR.

Here is a possible way for reviewing the changes:
* Start by looking at one or two of the bundled sample project, and read the documentation. Do you like the new api? Do you find it better than the old version? Is there anything more you'd like me to document?
* Review the actual play-slick code and tests together.

Remaining items in my TODO list are:

* Ship ~~0.9.0-M5~~ 1.0.0-RC1 (I've already updated the README, so I'd like to cut a release as soon as this PR gets merged).
* Find how the old and new documentation can coexist within the same repo (I see no reason for not keeping around the doc for v0.8).
* Help updating play-slick activator templates (e.g., https://github.com/loicdescotte/activator-play-slick and https://github.com/mkelland/activator-play-slick)
* Ticketing triage (several tickets should be closed when this PR is merged).

### CODE REVIEW

File: src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  def shutdown(): Unit = {
+    // no-op: shutting down dbs is automatically managed by `slickApi`
+    ()
+  }
+}
+
+private[slick] object DBApiAdapter {
+  private class DatabaseAdapter(_name: DbName, dbConfig: DatabaseConfig[JdbcProfile]) extends PlayDatabase {
+    def name: String = _name.value
+    def dataSource: DataSource = throw new UnsupportedOperationException
```
Comment (dotta) [inline src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala]: @szeiger Is there a way to return a `DataSource` from `DatabaseConfig[JdbcProfile]`?

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+The Play Slick module makes [Slick](http://slick.typesafe.com/) a first-class citizen of Play.
+
+The Play Slick module consists of 3 features:
+
+  - Integration of Slick into Play's application lifecycle. Because Slick now creates and manages both a connection pool and a thread pool, integrating Slick with Play boils down to ensure that all resources allocated by Slick are shutted down when a Play application is stopped (or reloaded).
+  - A DDL plugin, that reads Slick tables and automatically creates schema updates on reload. This is useful in particular for demos and to get started.
+  - Support for Play database evolutions.
+
+Play Slick currently supports Slick 3.0 with Play 2.4, for both Scala 2.10 and 2.11. If you need to use an older version of Play or Slick, have a look at the compatibility matrix in the [play-slick README] to know what version you should be using. Mind that the remainder of this guide explains how to use the latest release of the Play Slick, and will not be appropriate if you are using version 0.8 or lower.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md]: s/of the Play Slick/of Play Slick

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+Also, for the complete list of configuration parameters that you can use in your project's **application.conf**, see the Slick ScalaDoc for [Database.forConfig].
+
+[play-slick README]: https://github.com/playframework/play-slick#versioning
+[Slick documentation]: http://slick.typesafe.com/docs
+[Database.forConfig]: http://slick.typesafe.com/doc/3.0.0-RC3/api/index.html#slick.jdbc.JdbcBackend$DatabaseFactoryDef@forConfig(String,Config,Driver):Database
+
+## Usage
+
+After having properly configured your Slick databases, you can retrieve a `DatabaseConfig` (which is a Slick type bundling a database and driver) in two different ways. Either by using dependency injection, or through a global lookup via the `DatabaseConfigProvider` singleton.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md]: s/your Slick database/a Slick database

Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md]: s/retrieve/obtain

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+> Note: A Slick database instance manages a thread pool and a connection pool. In general, you should not need to shut down the database explicitly in your code (by calling its `close` method), as the Play Slick module takes care of this already.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md]: s/the database/a database

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+slick.evolutions.default="models.*"
+```
+
+It follows the same format as the Ebean plugin: `models.*` means all Tables in the models package should be run on the default database.
+
+It is possible to specify individual objects like:
+
+```conf
+slick.evolutions.default="models.Users,models.Settings"
+```
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlick.md]: Add a section about evolutions

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlickAdvancedTopics.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# Play Slick Advanced Topics
+
+## Connection Pool
+
+With Slick 3.0 release, Slick starts and controls both a connection pool and a thread pool for optimal asynchronous execution of your database actions.
+
+It's worth noting that while Slick allows using a different connection pool than [HikariCP] (though, Slick currently only offers built-in support for HikariCP, and requires you to provide an implementation of [JdbcDataSourceFactory] if you want to use a different connection pool), Play Slick currently doesn't allow using a different connection pool than HikariCP. If you find yourself needing this feature, please drop us a note on the [playframework-dev] mailing list.
+
+> Note: As stated in the [Slick documentation], reasonable defaults for the connection pool sizes are calculated from the thread pool size. Therefore, you should only need to tune `numThreads` and `queueSize` in most cases, for each of your database configurations.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlickAdvancedTopics.md]: s/connection pool sizes/connection pool size

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlickAdvancedTopics.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+> Note: Changing the value of `play.db.pool` won't affect what connection pool Slick is using. In fact, it's worth noting that Play Slick disables both the `play.api.db.DBModule` and `play.api.db.HikariCPModule` Play modules, and it will let Slick be in control of creating and managing the connection pool.
+
+## Thread Pool
+
+With Slick 3.0 release, Slick starts and controls both a thread pool and a connection pool for optimal asynchronous execution of your database actions.
+
+For optimal execution of code using the database, you may need to tune the `numThreads` and `queueSize` parameters, for each of your database configurations. Refer to the [Slick documentation] for details.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlickAdvancedTopics.md]: s/For optimal execution of code using the database/For optimal performance

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
+# Play Slick v0.9 Migration Guide
+
+This is a guide for migrating from Play Slick v0.8 to v0.9.
+
+It assumes you have already migrated your project to use Play 2.4 (see [Play 2.4 Migration Guide]), that you have read the [Slick 3 documentation], and are ready to migrate your Play application to use the new Slick Database I/O Actions API.
+
+## Build changes
+
+Update the Play Slick dependency in your sbt build to `0.9.0`:
+
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md]: Remove remainder of this section and add a link to PlaySlick#Setup

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+While the parameter `slick.db.execution.context` was used to name the thread pools created by Play Slick. In Slick 3, each thread pool is named using the Slick database configuration path, i.e., if in your **application.conf** you have provided a Slick configuration for the database named `default`, then Slick will create a thread pool named `default` for executing the database action on the default database. Note that the name used for the thread pool is not configurable.
+
+[Database.forConfig]: http://slick.typesafe.com/doc/3.0.0-RC3/api/index.html#slick.jdbc.JdbcBackend$DatabaseFactoryDef@forConfig(String,Config,Driver):Database
+
+## `Profile` was removed
+
+The trait `Profile` was removed and you can use instead `HasDatabaseConfigProvider` or `HasDatabaseConfig` with similar results.
+
+The trait to use depend on what approach you select to retrieve a Slick database and driver (i.e., an instance of `DatabaseConfig`). If you decide to use dependency injection, then `HasDatabaseConfigProvider` will serve you well. Otherwise, use `HasDatabaseConfig`.
```
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md]: This and the next section should be moved in the PlaySlick doc page (and here we should add a link to its content)

File: samples/iteratee/app/dao/RecordsDAO.scala [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    def byId(id: Rep[Int]) = q.filter(_.id === id)
+  }
+
+  def all(): Future[Seq[Record]] = db.run(records.result)
+  def count(): Future[Int] = db.run(records.length.result)
+  def insert(records: Seq[Record]): Future[Option[Int]] = db.run(this.records ++= records)
+
+  /** This is the interesting bit: enumerate the query for all Records in chunks of 2 */
+  def enumerateAllInChunksOfTwo: Enumerator[List[Record]] = {
+    import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (dotta) [inline samples/iteratee/app/dao/RecordsDAO.scala]: Change this to use Play Execution context

File: src/main/scala/play/api/db/slick/DatabaseConfigProvider.scala [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  * private class CatsTable(tag: Tag) extends Table[Cat](tag, "CAT") {
+  * def name = column[String]("NAME", O.PrimaryKey)
+  * def color = column[String]("COLOR")
+  * def * = (name, color) <> (Cat.tupled, Cat.unapply _)
+  * }
+  * }
+  * }}}
+  *
+  * Of course, you do not need to define a DAO to use this trait (the above it is really just an example of usage).
+  *
```
Comment (dotta) [inline src/main/scala/play/api/db/slick/DatabaseConfigProvider.scala]: remove this empty line

File: src/main/scala/play/api/db/slick/SlickApi.scala [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        val playConfig = PlayConfig(config)
+        playConfig.get[Map[String, Config]](slickDbKey)
+      } else Map.empty[String, Config]
+    }
+    (for ((name, config) <- configs) yield
+      (DbName(name), new DatabaseConfigFactory(name, config, lifecycle)))(collection.breakOut)
+  }
+
+  // Be careful that accessing this field will trigger initialization of ALL database configs!
+  private lazy val dbconfigByName: List[(DbName, DatabaseConfig[BasicProfile])] =
```
Comment (dotta) [inline src/main/scala/play/api/db/slick/SlickApi.scala]: `allDbConfigs` may be a better name

File: src/main/scala/play/api/db/slick/SlickApi.scala [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+import play.api.PlayConfig
+import play.api.PlayException
+import play.api.inject.ApplicationLifecycle
+
+import slick.backend.DatabaseConfig
+import slick.profile.BasicProfile
+
+trait SlickApi {
+  /** Returns all database configs, for all databases defined in the **application.conf**.
+   *  Throws a PlayException if the `default` database config cannot be created.
```
Comment (dotta) [inline src/main/scala/play/api/db/slick/SlickApi.scala]: > Throws a PlayException if the `default` database config cannot be created.

This is wrong :-)

File: src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala (hunk 1 above)
Comment (szeiger) [inline src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala]: You can look at `.db.source` but there's no common abstraction, so you'll have to handle `DataSourceJdbcDataSource` and `HikariCPJdbcDataSource` separately (both have a `ds` field). We could make the latter a subclass of the former in 3.1.

File: docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md (hunk 9 above)
Comment (dotta) [inline docs/manual/working/scalaGuide/main/sql/slick/PlaySlickMigrationGuide.md]: I've changed my mind on this one. So, things are good as they are.

File: src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala (hunk 1 above)
Comment (dotta) [inline src/main/scala/play/api/db/slick/internal/DBApiAdapter.scala]: Got it, thanks.

**Comment by misto** [comment]
:+1: looking forward to upgrading my applications!

**Comment by dotta** [comment]
@misto Awesome!

**Comment by iozozturk** [comment]
congratz :+1:

**Comment by dumb-questions** [comment]
Hi,

thanks for the release. :) How do I add this snapshot to my play 2.4.0-RC1 project? When I follow the instructions from the play docs I get an unresolved dependency for play-slick (I tried "com.typesafe.play" %% "play-slick" % "1.0.0-RC1" as well as "com.typesafe.play" %% "play-slick" % "1.0.0-SNAPSHOT")

[warn] com.typesafe.play#play-slick_2.11;1.0.0-SNAPSHOT: not found

As resolvers I use the following:

"Scalaz Bintray Repo" at "https://dl.bintray.com/scalaz/releases",
"Typesafe repository" at "https://repo.typesafe.com/typesafe/releases/",
"Typesafe Snapshots" at "https://repo.typesafe.com/typesafe/snapshots/",
"Sonatype OSS Snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"

**Pull request merged**